### PR TITLE
Use VK_EXT_swapchain_maintenance1 in the Swapchain Recreation sample

### PR DESCRIPTION
**Pull request by ShabbyX**
## Description

This change builds on https://github.com/KhronosGroup/Vulkan-Samples/pull/613 which adds a sample demonstrating safe recycling of present semaphores and efficient swapchain recreation. This change takes advantage of VK_EXT_swapchain_maintenance1 to perform the same job with less complication / more efficiently.

This sample has been tested on Nvidia/Linux/X11, as well as SwiftShader/Linux/X11.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/api/swapchain_recreation/swapchain_recreation.h
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.h]: let's use override and be more modern.  Remove virtual and replace it with override at the end and this will be good.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: In the above three lines there's a type mismatch.  Please consider using static_cast to change the uint32_t to float explicitly.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: old_swapchain shadows the argument for the function at line 815  Rename please

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: avoid using static_cast to downcast to a derived class.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: Please add a default to this switch statement to avoid having to define all potential enum values and to help keep warnings out of code.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: place a ',' before so for grammatical correctness.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: swap the 'e' and the 's' in present.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (gpx1000) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: triple (drop the extra 'p')

**Review by gpx1000 — CHANGES_REQUESTED** [review]

File: samples/api/swapchain_recreation/swapchain_recreation.h
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.h]: Done in parent PR

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done in parent PR

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Replied to in parent PR

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done in parent PR

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done in parent PR

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done in parent PR

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done

**Review by gpx1000 — DISMISSED** [review]

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (JoseEmilio-ARM) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: thanks for the sample! looks good to me. Would you have a chance to test it on Android? maybe not all functionality like these controls need to be there, but could be useful to test the extension on that platform.

**Comment by marty-johnson59** [comment]
Hi @ShabbyX, we briefly discussed this on the Samples call today and had questions on timing.  Do you know if (or when) there are HW drivers that support this?  It looked like it is available on Swiftshader and possibly RADV.  We're thinking we should hold off until more vendors support this in drivers?  We would be interested in your thoughts on timing for this sample.  Thanks.

**Comment by ShabbyX** [comment]
> We're thinking we should hold off until more vendors support this in drivers?

Yes, that was my understanding in the call a few weeks ago, which is why I split the PRs. I am in no rush to land this, so take your time!

> Do you know if (or when) there are HW drivers that support this?

I know Hans-Kristian had an implementation for mesa (I believe that'd be all mesa drivers), but don't know if that's merged. I also know that Nvidia is working on it, but of course release timing depends on them.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: I'm too swamped right now, but maybe one day. The extension _is_ tested on Android both through CTS and ANGLE if you are more worried about the extension itself (and less about the sample)

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (asuessenbach) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp:54]: If you move the call to `check_for_maintenance1` to after `VulkanSample::prepare`, you don't have to `volkInitialize`.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (asuessenbach) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp]: Maybe use std::find_if to look for the extension?

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (asuessenbach) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp:112]: This function is only called if `has_maintenance1` is true -> no need to check it here again.

**Review by asuessenbach — COMMENTED** [review]
Unfortunately, I don't have HW around supporting VK_EXT_surface_maintenance1.
But the changes compared to #613 look pretty clear.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp:54]: Nice, done

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Done

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp:112]: Moved the call up to before the if instead.

I think it's best if applications did it this way, they'd get a list of compatible present modes which would be one-sized if maint1 is not supported. Then when checking if a present mode change is necessary, they don't have to bother with whether maint1 is supported or not.

I cleaned up `are_present_modes_compatible` accordingly to only look at the compatible present modes (no current == desired or extension check). Also added statistics to be printed in the end about how many swapchains were created during the lifetime of the sample.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp:54]: Actually no, the reason this has to be done before `VulkanSample::prepare` is that it calls `add_instance_extension`, and `VulkanSample::prepare` creates the instance. Calling this after effectively means the extension is not enabled.

**Comment by gary-sweet** [comment]
I don't have a driver that supports swapchain_maintenance1 yet, so all I can do is confirm that this doesn't break the existing functionality for me.

**Review by asuessenbach — DISMISSED** [review]

**Review by SaschaWillems — COMMENTED** [review]
Works fine for me. 

Will approve once the merge conflicts have been resolved :)

**Comment by ShabbyX** [comment]
> Works fine for me.
> 
> Will approve once the merge conflicts have been resolved :)

Done

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (asuessenbach) [inline samples/api/swapchain_recreation/swapchain_recreation.cpp:79]: I think, the way to handle the extensions would be to unconditionally add them in the `SwapchainRecreation` constructor, but mark them as optional.
Then, you don't need to initialize volk on your own, and can handle the implications of that extension not being there in `prepare`.

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp:79]: I'll try when I find the time, but I think the issue was with `surface_maintenance1` being an _instance_ extension that made it complicated.

File: samples/api/swapchain_recreation/swapchain_recreation.cpp
Comment (marty-johnson59) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Hi @ShabbyX, is this sample still in the plans?  Thanks

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Yes, it was ready a year ago IMO. The TSG decided to wait for a driver release to test this with, and surprisingly that hasn't happened yet. I know @HansKristian-Work had an implementation for mesa last year, but that never landed? Nvidia surprisingly hasn't released a driver yet either.

So, I guess it'll just have to continue to wait. If an Android-savvy person likes to try this sample on Android (like Jose Emilion? :D) and give it the green light, by all means go ahead!

Comment (SaschaWillems) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Looking at my database, AMD and Nvidia support the extension:
https://vulkan.gpuinfo.org/listdevicescoverage.php?extension=VK_EXT_swapchain_maintenance1&platform=all

AMD is currently Linux only, nvidia is Windows and Linux :)

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Huh, I checked there, how did I miss that?

Ok well, like I said last year the TSG decided to wait until someone (other than myself) can try this sample on a desktop driver. So by all means, please do so it can land.

Comment (marty-johnson59) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: :)  Excellent..

Comment (HansKristian-Work) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Mesa supports Wayland and X11. X11 backend recently received support for present mode toggling as well.

NV's prop driver has some bugs with swapchain fence though, at least on Linux. A fix is supposedly on its way.

Comment (marty-johnson59) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: OK, thanks.  We'll check it out and hopefully be close to merging by the next samples call..

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Just rebased, there was a minor conflict about .md->.adoc transition.

Also, running on SwiftShader right now, I see a new VVL error about the ANI semaphore not being unused. It triggers only once though, I'll have to look into it.

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: Could be a VVL bug? As far as my logs show, the semaphore is not in use:

```
[info] ------------------------------------
[info] USAGE:
[info]  - Press v to enable v-sync (default)
[info]  - Press n to disable v-sync
[info]  - Press p to enable switching between compatible present modes (default)
[info]  - Press r to disable switching between compatible present modes
[info] Set environment variable USE_MAINTENANCE1=no to avoid VK_EXT_surface_maintenance1
[info] ------------------------------------
[warning] =======
[warning] NEW FRAME semaphore 0xa000000000a, fence 0x90000000009, present sem: 0xb000000000b
[warning] Calling ANI with semaphore 0xa000000000a
[warning] Submit, wait on 0xa000000000a, signal 0xb000000000b, signal fence 0x90000000009
[warning] Present, wait on 0xb000000000b
[warning] Present, signal fence 0xf000000000f
[warning] Clean up present fence -- fence 0xf000000000f is signaled
[warning] Clean up present fence -- recycle present semaphore 0xb000000000b
[warning] =======
[warning] NEW FRAME semaphore 0xb000000000b, fence 0xf000000000f, present sem: 0x100000000010
[warning] Calling ANI with semaphore 0xb000000000b
[error] [framework/core/instance.cpp:50] 1461184347 - VUID-vkAcquireNextImageKHR-semaphore-01779: Validation Error: [ VUID-vkAcquireNextImageKHR-semaphore-01779 ] Object 0: handle = 0xb000000000b, type = VK_OBJECT_TYPE_SEMAPHORE; | MessageID = 0x5717e75b | vkAcquireNextImageKHR():  Semaphore must not have any pending operations. The Vulkan spec states: If semaphore is not VK_NULL_HANDLE it must not have any uncompleted signal or wait operations pending (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkAcquireNextImageKHR-semaphore-01779)
[warning] Submit, wait on 0xb000000000b, signal 0x100000000010, signal fence 0xf000000000f
[warning] Present, wait on 0x100000000010
[warning] Present, signal fence 0x120000000012
[warning] Clean up present fence -- fence 0x120000000012 is signaled
[warning] Clean up present fence -- recycle present semaphore 0x100000000010
[warning] Recycle semaphore 0xa000000000a, fence 0x90000000009 (waited)
```

It's warning about the second call to ANI, which is using the semaphore used in the previous present operation. However, the fence for that semaphore is already signaled per "[warning] Clean up present fence -- fence 0xf000000000f is signaled". Maybe VVL is not accounting for VK_EXT_swapchain_maintenance1. I don't believe the app is doing anything wrong here. Diff used to reproduce these logs:

```
diff --git a/samples/api/swapchain_recreation/swapchain_recreation.cpp b/samples/api/swapchain_recreation/swapchain_recreation.cpp
index 1a5b29c..70c9e19 100644
--- a/samples/api/swapchain_recreation/swapchain_recreation.cpp
+++ b/samples/api/swapchain_recreation/swapchain_recreation.cpp
@@ -418,6 +418,7 @@ void SwapchainRecreation::setup_frame()
                vkWaitForFences(get_device_handle(), 1, &frame.submit_fence, true, UINT64_MAX);
 
                // Reset/recycle resources, they are no longer in use.
+               LOGW("Recycle semaphore {}, fence {} (waited)", (void *)frame.acquire_semaphore, (void *)frame.submit_fence);
                recycle_fence(frame.submit_fence);
                recycle_semaphore(frame.acquire_semaphore);
                vkResetCommandPool(get_device_handle(), frame.command_pool, 0);
@@ -443,6 +444,8 @@ void SwapchainRecreation::setup_frame()
        frame.submit_fence      = get_fence();
        frame.acquire_semaphore = get_semaphore();
        frame.present_semaphore = get_semaphore();
+       LOGW("=======");
+       LOGW("NEW FRAME semaphore {}, fence {}, present sem: {}", (void *)frame.acquire_semaphore, (void *)frame.submit_fence, (void *)frame.present_semaphore);
 
        if (frame.command_pool == VK_NULL_HANDLE)
        {
@@ -542,6 +545,7 @@ void SwapchainRecreation::render(uint32_t index)
        info.pWaitDstStageMask    = &wait_stage;
        info.signalSemaphoreCount = 1;
        info.pSignalSemaphores    = &frame.present_semaphore;
+       LOGW("Submit, wait on {}, signal {}, signal fence {}", (void *)frame.acquire_semaphore, (void *)frame.present_semaphore, (void *)frame.submit_fence);
        VK_CHECK(vkQueueSubmit(queue->get_handle(), 1, &info, frame.submit_fence));
 }
 
@@ -576,6 +580,7 @@ VkResult SwapchainRecreation::acquire_next_image(uint32_t *index)
        // There is no need for this with VK_EXT_swapchain_maintenance1.
        VkFence acquire_fence = has_maintenance1 ? VK_NULL_HANDLE : get_fence();
 
+       LOGW("Calling ANI with semaphore {}", (void *)frame.acquire_semaphore);
        VkResult result = vkAcquireNextImageKHR(get_device_handle(), swapchain, UINT64_MAX, frame.acquire_semaphore, acquire_fence, index);
 
        if (has_maintenance1 && (result == VK_SUCCESS || result == VK_SUBOPTIMAL_KHR))
@@ -623,6 +628,7 @@ VkResult SwapchainRecreation::present_image(uint32_t index)
        present.pImageIndices      = &index;
        present.waitSemaphoreCount = 1;
        present.pWaitSemaphores    = &frame.present_semaphore;
+       LOGW("Present, wait on {}", (void *)frame.present_semaphore);
 
        // When VK_EXT_swapchain_maintenance1 is enabled, add a fence to the present operation,
        // which is signaled when the resources associated with present operation can be freed.
@@ -635,6 +641,7 @@ VkResult SwapchainRecreation::present_image(uint32_t index)
 
                fence_info.swapchainCount = 1;
                fence_info.pFences        = &present_fence;
+               LOGW("Present, signal fence {}", (void *)present_fence);
 
                present.pNext = &fence_info;
 
@@ -710,6 +717,7 @@ void SwapchainRecreation::cleanup_present_history()
                }
                VK_CHECK(result);
 
+               LOGW("Clean up present fence -- fence {} is signaled", (void *)present_info.cleanup_fence);
                cleanup_present_info(present_info);
                present_history.pop_front();
        }
@@ -752,6 +760,7 @@ void SwapchainRecreation::cleanup_present_info(PresentOperationInfo &present_inf
        // clean up.  That fence is placed in the present history just for clean up purposes.
        if (present_info.present_semaphore != VK_NULL_HANDLE)
        {
+               LOGW("Clean up present fence -- recycle present semaphore {}", (void *)present_info.present_semaphore);
                recycle_semaphore(present_info.present_semaphore);
        }
```

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: I noticed in the meeting notes that this sample is "Waiting on updates". There are no updates coming, this was confirmed to be a VVL gap (https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/7025). Unless you are waiting for the VVL error to go away, that's blocked on that issue. If not, the change is good to land.

Comment (ShabbyX) [reply samples/api/swapchain_recreation/swapchain_recreation.cpp]: The VVL bug has been fixed. I just rebased this to retrigger the CI.

**Review by SaschaWillems — APPROVED** [review]
Works fine on my NVIDIA RTX 4070 on Windows 11, which finally supports the extension.

I can also confirm that the validation error mentioned above is gone with a build of the latest validation layers from source.

**Review by gpx1000 — APPROVED** [review]

**Pull request merged**
